tensorflow/python/autograph/operators/logical_test.py

```diff
@@ -83,6 +85,18 @@ def test_not_tf(self):
       t = logical.not_(self._tf_false())
       self.assertEqual(self.evaluate(t), True)
 
+  # Test case for GitHub issue 40471
```

Comment:
Since the fix itself is not directly related to autograph, it would be best to have this test somewhere in a location that's closer to the `tf.equal` op.

I think `math_ops_test.cc` is a good place to verify this logic, since it already includes a basic test. Consider adding a second test which has a more complex shape, like the ones tested here. Feel free to add a new test case (e.g. `TEST(MathOpsTest, EqualOp)`).

Optionally, there is also `tensorflow/python/kernel_tests`, which could be expanded by adding an `equal_test.py` to contain a few more test cases, which should include:
 * matching shapes
 * incompatible shapes
 * dynamic, matching shapes
 * dynamic, incompatible shapes (this last one is most important to make sure that op kernels weren't relying on the shape inference to behave correctly)